src/common/Throttle.h

```diff
@@ -330,4 +332,84 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+
+class TokenBucketThrottle {
+
+  class Bucket {
+    CephContext *cct;
+    const std::string name;
+    std::atomic<int64_t> remain = { 0 }, max = { 0 };
+    Mutex lock;
+    list<Cond*> cond;
+    int64_t get(int64_t c);
+    friend class TokenBucketThrottle;
+
+  public:
+    Bucket(CephContext *cct, const std::string& n, int64_t m = 0);
+    ~Bucket();
+
+  public:
+    void set_max(int64_t m);
+    bool get(int64_t c, Context *ctx);
+    int64_t put(int64_t c);
+  };
+
+  class Blocker {
+    int64_t token_requested;
+    Context *got_token_ctx;
+
+    Blocker(int64_t _token_requested, Context *ctx)
+    : token_requested(_token_requested), got_token_ctx(ctx) {}
+
+    friend class TokenBucketThrottle;
+  };
+
+  CephContext *m_cct;
+  Bucket m_throttle;
+  int64_t m_avg;
+  SafeTimer *m_timer;
+  Mutex *m_timer_lock;
+  FunctionContext *m_token_ctx;
+  list<Blocker*> m_blockers;
```

Comment:
Why use heap allocation?

Reply:
Follow the style of ```std::list<T *> m_items;``` in PointerWQ

Reply:
That was due to the need to support inherited classes -- this class owns ```Blocker``` so there is no need. Just use ```emplace``` or move semantics